content/en/blog/_posts/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm.md

```diff
@@ -0,0 +1,204 @@
+
+With their help you can deploy a fully working Kubernetes cluster inside another Kubernetes using Helm just in a couple of commands. How and why?
+
+Let me introduce you how our infrastructure works. All our physical servers can be divided into two groups: **control-plane** and **compute** nodes. Control plane nodes are usually set up manually, have a stable OS installed, and designed to run all cluster services, including Kubernetes control-plane. The main task of these nodes is to ensure the smooth operation of the cluster itself. Compute nodes do not have any operating system installed by default, Instead they are booting the OS image over the network directly from the control plane nodes. Their work is to carry out the workload.
+
+![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/scheme01.jpg)
+
+Once nodes have downloaded their image, they can continue to work without keeping connection to the PXE server. That is, a PXE server is just keeping rootfs image and does not hold any other complex logic. After our nodes have booted, we can safely restart the PXE server, nothing critical will happen to them.
+
+![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/scheme02.jpg)
+
+After booting, the first thing our nodes do is join to the existing Kubernetes cluster, namely, execute the **kubeadm join** command. So, kube-scheduler could schedule some pods on them and launch various workloads afterwards. From the begining we used the scheme when nodes were joined into the same cluster used for the control-plane nodes.
+
+![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/scheme03.jpg)
+
+This scheme worked stably for over two years. However later we decided to add containerized Kubernetes to it. And now we can spawn new Kubernetes-clusters very easely right on our control-plane nodes which are now is members special admin-cluster. And compute nodes now can be joined directly to their own clusters - depending on the configuration.
+
+![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/scheme04.jpg)
+
+## Kubefarm
+
+This project came about with the goal to make anyone could deploy such an infrastructure just in a couple of commands using Helm and get about the same in the end.
+
+At this time, we moved away from the idea of a monocluster. Because it turned out to be not very convenient for managing several development teams work in the same cluster. The fact is that Kubernetes was never designed as a multi-tenant solution and at the moment it does not provide sufficient means of isolation between projects. Therefore, running separate clusters for each team turned out to be a good idea. However, there should not be too many clusters to let them be convenient to manage. Nor is it too small to have sufficient independence between development teams.
```

Comment:
```suggestion
At this time, we moved away from the idea of a monocluster. Because it turned out to be not very convenient for managing work of several development teams in the same cluster. The fact is that Kubernetes was never designed as a multi-tenant solution and at the moment it does not provide sufficient means of isolation between projects. Therefore, running separate clusters for each team turned out to be a good idea. However, there should not be too many clusters, to let them be convenient to manage. Nor is it too small to have sufficient independence between development teams.
```